From 83ee53a2f7f4f3c92ca8d0a18ea2d5e94fb69bc1 Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Thu, 19 Sep 2024 17:41:32 +0800
Subject: [PATCH 22] add dummy ijkbluray2 protocol

---
 libavformat/ijkutils.c  | 1 +
 libavformat/protocols.c | 1 +
 2 files changed, 2 insertions(+)

diff --git a/libavformat/ijkutils.c b/libavformat/ijkutils.c
index 25faad1..12f2da7 100644
--- a/libavformat/ijkutils.c
+++ b/libavformat/ijkutils.c
@@ -65,6 +65,7 @@ IJK_DUMMY_PROTOCOL(ijkhttphook);
 IJK_DUMMY_PROTOCOL(ijksegment);
 IJK_DUMMY_PROTOCOL(ijktcphook);
 IJK_DUMMY_PROTOCOL(ijkio);
+IJK_DUMMY_PROTOCOL(ijkbluray2);
 
 #define IJK_FF_DEMUXER(x)                                                                          \
 extern AVInputFormat ff_##x##_demuxer;                                                               \
diff --git a/libavformat/protocols.c b/libavformat/protocols.c
index 73df344..8819ecb 100644
--- a/libavformat/protocols.c
+++ b/libavformat/protocols.c
@@ -83,6 +83,7 @@ extern const URLProtocol ff_ijkmediadatasource_protocol;
 extern const URLProtocol ff_ijksegment_protocol;
 extern const URLProtocol ff_ijktcphook_protocol;
 extern const URLProtocol ff_ijkio_protocol;
+extern const URLProtocol ff_ijkbluray2_protocol;
 
 extern const URLProtocol ff_dvd_protocol;
 
-- 
2.39.3 (Apple Git-146)

