From b5176510c15322f2223c5e16af20e65be98db163 Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Wed, 19 Mar 2025 13:22:27 +0800
Subject: [PATCH 29] bluray protocol add dvd fallback

---
 libavformat/demux.c | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/libavformat/demux.c b/libavformat/demux.c
index 51fff7b..05e02db 100644
--- a/libavformat/demux.c
+++ b/libavformat/demux.c
@@ -158,7 +158,10 @@ static int init_input(AVFormatContext *s, const char *filename,
     int ret;
     AVProbeData pd = { filename, NULL, 0 };
     int score = AVPROBE_SCORE_RETRY;
-
+    AVDictionary *tmp_opts = NULL;
+    if (options && (av_stristart(filename, "bluray://", NULL) || av_stristart(filename, "dvd://", NULL))) {
+        av_dict_copy(&tmp_opts, *options, 0);
+    }
     if (s->pb) {
         s->flags |= AVFMT_FLAG_CUSTOM_IO;
         if (!s->iformat)
@@ -174,8 +177,23 @@ static int init_input(AVFormatContext *s, const char *filename,
         (!s->iformat && (s->iformat = av_probe_input_format2(&pd, 0, &score))))
         return score;
 
-    if ((ret = s->io_open(s, &s->pb, filename, AVIO_FLAG_READ | s->avio_flags, options)) < 0)
-        return ret;
+    if ((ret = s->io_open(s, &s->pb, filename, AVIO_FLAG_READ | s->avio_flags,
+                          options)) < 0) {
+        if (av_stristart(filename, "bluray://", NULL)) {
+            const char *dvd_name =
+                av_strireplace(filename, "bluray://", "dvd://");
+            ret = init_input(s, dvd_name, &tmp_opts);
+            av_dict_free(&tmp_opts);
+            return ret;
+        } else if (av_stristart(filename, "dvd://", NULL)) {
+            const char *a_name = av_strireplace(filename, "dvd://", "");
+            ret = init_input(s, a_name, &tmp_opts);
+            av_dict_free(&tmp_opts);
+            return ret;
+        } else {
+            return ret;
+        }
+    }
 
     if (s->iformat)
         return 0;
-- 
2.39.5 (Apple Git-154)

