From 3f0966825de7bf6ff6cb858c70f197fbeafe3419 Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Thu, 17 Apr 2025 10:36:54 +0800
Subject: [PATCH 31] let avio can retry when avio encounter error but not eof

---
 libavformat/aviobuf.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/libavformat/aviobuf.c b/libavformat/aviobuf.c
index 2899c75..06012b7 100644
--- a/libavformat/aviobuf.c
+++ b/libavformat/aviobuf.c
@@ -569,7 +569,10 @@ static void fill_buffer(AVIOContext *s)
         s->eof_reached = 1;
 
     /* no need to do anything if EOF already reached */
-    if (s->eof_reached)
+    /* when "len = read_packet_wrapper(s, dst, len);" has error, eof_reached marked to 1
+     * in effect the stream not eof but encounter network error
+     */
+    if (s->eof_reached && !s->error)
         return;
 
     if (s->update_checksum && dst == s->buffer) {
-- 
2.39.5 (Apple Git-154)

