From 7b54df2c93d327611d594f9b4ea1475bcf729cb6 Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Fri, 29 Jul 2022 12:54:09 +0800
Subject: [PATCH 04/15] file use file_seek

---
 libavformat/file.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libavformat/file.c b/libavformat/file.c
index 98c9e81..edd939c 100644
--- a/libavformat/file.c
+++ b/libavformat/file.c
@@ -394,7 +394,7 @@ static int pipe_open(URLContext *h, const char *filename, int flags)
     setmode(fd, O_BINARY);
 #endif
     c->fd = fd;
-    h->is_streamed = 1;
+    h->is_streamed = 0;
     return 0;
 }
 
@@ -403,6 +403,7 @@ const URLProtocol ff_pipe_protocol = {
     .url_open            = pipe_open,
     .url_read            = file_read,
     .url_write           = file_write,
+    .url_seek            = file_seek,
     .url_get_file_handle = file_get_handle,
     .url_check           = file_check,
     .priv_data_size      = sizeof(FileContext),
-- 
2.39.3 (Apple Git-145)

