From 7fc90c789db0167482da8dc300e7fc496109f44f Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Wed, 13 Dec 2023 11:05:39 +0800
Subject: [PATCH 14/15] record bluray debug log

---
 libavformat/bluray.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/libavformat/bluray.c b/libavformat/bluray.c
index 644166f..8814dc2 100644
--- a/libavformat/bluray.c
+++ b/libavformat/bluray.c
@@ -114,8 +114,24 @@ static int bluray_close(URLContext *h)
     return 0;
 }
 
+#ifdef DEBUG_BLURAY
+#include <libbluray/log_control.h>
+#define BLURAY_DEBUG_MASK 0xFFFFF //(0xFFFFF & ~DBG_STREAM)
+
+static void bluray_DebugHandler(const char *psz)
+{
+    size_t len = strlen(psz);
+    if(len < 1) return;
+    av_log(NULL, AV_LOG_DEBUG, "[bluray] %s\n",psz);
+}
+#endif
+
 static int bluray_open(URLContext *h, const char *path, int flags)
 {
+#ifdef DEBUG_BLURAY
+    bd_set_debug_mask(BLURAY_DEBUG_MASK);
+    bd_set_debug_handler(bluray_DebugHandler);
+#endif
     BlurayContext *bd = h->priv_data;
     int num_title_idx;
     const char *diskname = path;
-- 
2.39.3 (Apple Git-145)

