From 07fe7d7b1450b07be3e3234485283d1d8c1dbbfe Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Tue, 25 Aug 2026 11:46:34 +0800
Subject: [PATCH 3/3] adapt to iOS and tvOS platform

---
 meson.build     | 8 +++++++-
 src/meson.build | 6 +++++-
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/meson.build b/meson.build
index 9e5e897..ad1eadd 100644
--- a/meson.build
+++ b/meson.build
@@ -54,7 +54,13 @@ endif
 
 if host_machine.system() == 'darwin'
     extra_dependencies += dependency('appleframeworks',
-        modules: ['CoreFoundation', 'DiskArbitration'])
+        modules: ['CoreFoundation'])
+    da_dep = dependency('appleframeworks',
+        modules: ['DiskArbitration'], required: false)
+    if da_dep.found()
+        extra_dependencies += da_dep
+        cdata.set('HAVE_DISKARBITRATION', 1)
+    endif
 elif host_machine.system() == 'netbsd'
     test_args += '-D_NETBSD_SOURCE'
 elif host_machine.system() == 'openbsd'
diff --git a/src/meson.build b/src/meson.build
index 255fbc4..03386bb 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -77,8 +77,12 @@ elif host_machine.system() == 'darwin'
         'file/dirs_darwin.c',
         'file/dl_posix.c',
         'file/file_posix.c',
-        'file/mount_darwin.c',
     )
+    if cdata.has('HAVE_DISKARBITRATION')
+        libbluray_src += files('file/mount_darwin.c')
+    else
+        libbluray_src += files('file/mount.c')
+    endif
 else
     libbluray_src += files(
         'file/dir_posix.c',
-- 
2.50.1 (Apple Git-155)

