From e294f9626d50ce1808253f0ba28c4a60e6c78f0e Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Tue, 12 May 2026 09:46:53 +0800
Subject: [PATCH] add 3 dummy ijkhttp protocols

---
 libavformat/ijkutils.c  | 3 +++
 libavformat/protocols.c | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/libavformat/ijkutils.c b/libavformat/ijkutils.c
index bc851d2914..31e1e8064b 100644
--- a/libavformat/ijkutils.c
+++ b/libavformat/ijkutils.c
@@ -69,6 +69,9 @@ IJK_DUMMY_PROTOCOL(ijkhttphook);
 IJK_DUMMY_PROTOCOL(ijksegment);
 IJK_DUMMY_PROTOCOL(ijktcphook);
 IJK_DUMMY_PROTOCOL(ijkio);
+IJK_DUMMY_PROTOCOL(ijkhttp1);
+IJK_DUMMY_PROTOCOL(ijkhttp2);
+IJK_DUMMY_PROTOCOL(ijkhttp3);
 
 #define IJK_FF_DEMUXER(x)                                                                          \
 extern FFInputFormat ff_##x##_demuxer;                                                               \
diff --git a/libavformat/protocols.c b/libavformat/protocols.c
index 719caf8492..b291b9f97c 100644
--- a/libavformat/protocols.c
+++ b/libavformat/protocols.c
@@ -84,6 +84,9 @@ extern const URLProtocol ff_ijkmediadatasource_protocol;
 extern const URLProtocol ff_ijksegment_protocol;
 extern const URLProtocol ff_ijktcphook_protocol;
 extern const URLProtocol ff_ijkio_protocol;
+extern const URLProtocol ff_ijkhttp1_protocol;
+extern const URLProtocol ff_ijkhttp2_protocol;
+extern const URLProtocol ff_ijkhttp3_protocol;
 
 #include "libavformat/protocol_list.c"
 
-- 
2.50.1 (Apple Git-155)

