From 30f20ff3ec0de99847be4da379f497bbacc61501 Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Wed, 28 May 2025 18:39:38 +0800
Subject: [PATCH] fix lrcdec read line bug on osx

---
 libavformat/lrcdec.c | 20 +++++++++++---------
 1 file changed, 11 insertions(+), 9 deletions(-)

diff --git a/libavformat/lrcdec.c b/libavformat/lrcdec.c
index 68c44bc..2bea1e1 100644
--- a/libavformat/lrcdec.c
+++ b/libavformat/lrcdec.c
@@ -105,15 +105,17 @@ static int64_t read_line(AVBPrint *buf, AVIOContext *pb)
     int64_t pos = avio_tell(pb);
 
     av_bprint_clear(buf);
-    while(!avio_feof(pb)) {
-        int c = avio_r8(pb);
-        if(c != '\r') {
-            av_bprint_chars(buf, c, 1);
-        }
-        if(c == '\n') {
-            break;
-        }
-    }
+    ff_read_line_to_bprint_overwrite(pb, buf);
+
+    // while(!avio_feof(pb)) {
+    //     int c = avio_r8(pb);
+    //     if(c != '\r') {
+    //         av_bprint_chars(buf, c, 1);
+    //     }
+    //     if(c == '\n') {
+    //         break;
+    //     }
+    // }
     return pos;
 }
 
-- 
2.39.5 (Apple Git-154)

