From 55f90bbd61881edaef80f0751264b8ca4d0cec43 Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Mon, 8 Dec 2025 12:48:09 +0800
Subject: [PATCH] fix build error

---
 CMakeLists.txt | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9abfa74..b0feb35 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,8 +3,9 @@
 # Run with -DTEST=ON to build unit tests
 
 PROJECT ( YUV C CXX )	# "C" is required even for C++ projects
-CMAKE_MINIMUM_REQUIRED( VERSION 2.8.12 )
+CMAKE_MINIMUM_REQUIRED( VERSION 3.10)
 OPTION( UNIT_TEST "Built unit tests" OFF )
+OPTION( BINARY "Built binary" OFF )
 
 SET ( ly_base_dir	${PROJECT_SOURCE_DIR} )
 SET ( ly_src_dir	${ly_base_dir}/source )
@@ -29,6 +30,7 @@ endif()
 # this creates the static library (.a)
 ADD_LIBRARY				( ${ly_lib_static} STATIC ${ly_source_files} )
 
+if(BUILD_SHARED_LIBS)
 # this creates the shared library (.so)
 ADD_LIBRARY				( ${ly_lib_shared} SHARED ${ly_source_files} )
 SET_TARGET_PROPERTIES	( ${ly_lib_shared} PROPERTIES OUTPUT_NAME "${ly_lib_name}" )
@@ -36,7 +38,9 @@ SET_TARGET_PROPERTIES	( ${ly_lib_shared} PROPERTIES PREFIX "lib" )
 if(WIN32)
   SET_TARGET_PROPERTIES	( ${ly_lib_shared} PROPERTIES IMPORT_PREFIX "lib" )
 endif()
+endif()
 
+if(BINARY)
 # this creates the cpuid tool
 ADD_EXECUTABLE      ( cpuid ${ly_base_dir}/util/cpuid.c )
 TARGET_LINK_LIBRARIES  ( cpuid ${ly_lib_static} )
@@ -55,6 +59,7 @@ if (JPEG_FOUND)
   target_link_libraries( ${ly_lib_shared} ${JPEG_LIBRARY} )
   add_definitions( -DHAVE_JPEG )
 endif()
+endif()
 
 if(UNIT_TEST)
   find_library(GTEST_LIBRARY gtest)
@@ -97,12 +102,16 @@ if(UNIT_TEST)
 endif()
 
 
+if(BINARY)
 # install the conversion tool, .so, .a, and all the header files
 INSTALL ( PROGRAMS ${CMAKE_BINARY_DIR}/yuvconvert			DESTINATION bin )
+endif()
 INSTALL ( TARGETS ${ly_lib_static}						DESTINATION lib )
+if(BUILD_SHARED_LIBS)
 INSTALL ( TARGETS ${ly_lib_shared} LIBRARY				DESTINATION lib RUNTIME DESTINATION bin )
+endif()
 INSTALL ( DIRECTORY ${PROJECT_SOURCE_DIR}/include/		DESTINATION include )
 
 # create the .deb and .rpm packages using cpack
-INCLUDE ( CM_linux_packages.cmake )
+# INCLUDE ( CM_linux_packages.cmake )
 
-- 
2.50.1 (Apple Git-155)

